// @vicinae/api/dist/utils/vicinae.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VicinaeClient = void 0;
const node_child_process_1 = require("node:child_process");
class VicinaeClient {
    options;
    constructor(options = { binaryPath: "vicinae" }) {
        this.options = options;
    }
    ping() {
        return this.invoke("/ping");
    }
    refreshDevSession(extensionId) {
        return this.invoke(`/api/extensions/develop/refresh?id=${extensionId}`);
    }
    startDevSession(extensionId) {
        return this.invoke(`/api/extensions/develop/start?id=${extensionId}`);
    }
    stopDevSession(extensionId) {
        return this.invoke(`/api/extensions/develop/stop?id=${extensionId}`);
    }
    invoke(endpoint) {
        if (endpoint.startsWith("/")) {
            endpoint = endpoint.slice(1);
        }
        const url = `vicinae://${endpoint}`;
        const result = (0, node_child_process_1.spawnSync)(this.options.binaryPath, [url]);
        if (result.error)
            return result.error;
        return result.status === 0 ? null : new Error(result.stderr.toString());
    }
}
exports.VicinaeClient = VicinaeClient;
