// @vicinae/api/dist/api/alert.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.confirmAlert = exports.Alert = void 0;
const bus_1 = require("./bus");
const ui_1 = require("./proto/ui");
var Alert;
(function (Alert) {
    let ActionStyle;
    (function (ActionStyle) {
        ActionStyle["Default"] = "default";
        ActionStyle["Destructive"] = "destructive";
        ActionStyle["Cancel"] = "cancel";
    })(ActionStyle = Alert.ActionStyle || (Alert.ActionStyle = {}));
})(Alert || (exports.Alert = Alert = {}));
const styleMap = {
    [Alert.ActionStyle.Default]: ui_1.ConfirmAlertActionStyle.Default,
    [Alert.ActionStyle.Destructive]: ui_1.ConfirmAlertActionStyle.Destructive,
    [Alert.ActionStyle.Cancel]: ui_1.ConfirmAlertActionStyle.Cancel,
};
const confirmAlert = async (options) => {
    return new Promise(async (resolve) => {
        const req = ui_1.ConfirmAlertRequest.create({
            title: options.title,
            description: options.message ?? "Are you sure?",
            rememberUserChoice: false,
            primaryAction: {
                title: options.primaryAction?.title ?? "Confirm",
                style: styleMap[options.primaryAction?.style ?? Alert.ActionStyle.Default],
            },
            dismissAction: {
                title: options.dismissAction?.title ?? "Cancel",
                style: styleMap[options.dismissAction?.style ?? Alert.ActionStyle.Cancel],
            },
        });
        const res = await bus_1.bus.turboRequest("ui.confirmAlert", req);
        if (!res.ok)
            return false;
        if (res.value.confirmed) {
            options.primaryAction?.onAction?.();
        }
        else {
            options.dismissAction?.onAction?.();
        }
        resolve(res.value.confirmed);
    });
};
exports.confirmAlert = confirmAlert;
