// @vicinae/api/dist/api/components/list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.List = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const crypto_1 = require("crypto");
const metadata_1 = require("./metadata");
const empty_view_1 = require("./empty-view");
const dropdown_1 = require("./dropdown");
const ListRoot = ({ searchBarAccessory, children, actions, ...props }) => {
    if (typeof props.enableFiltering === "boolean" &&
        typeof props.filtering === "undefined") {
        props.filtering = props.enableFiltering;
    }
    return ((0, jsx_runtime_1.jsxs)("list", { ...props, children: [searchBarAccessory, children, actions] }));
};
const ListItem = ({ detail, actions, ...props }) => {
    const id = (0, react_1.useRef)(props.id ?? (0, crypto_1.randomUUID)());
    return ((0, jsx_runtime_1.jsxs)("list-item", { ...props, id: id.current, children: [detail, actions] }));
};
const ListItemDetail = ({ metadata, ...props }) => {
    return (0, jsx_runtime_1.jsx)("list-item-detail", { ...props, children: metadata });
};
const ListSection = (props) => {
    return (0, jsx_runtime_1.jsx)("list-section", { ...props });
};
exports.List = Object.assign(ListRoot, {
    Section: ListSection,
    EmptyView: empty_view_1.EmptyView,
    Dropdown: dropdown_1.Dropdown,
    Item: Object.assign(ListItem, {
        Detail: Object.assign(ListItemDetail, {
            Metadata: metadata_1.Metadata,
        }),
    }),
});
