// @vicinae/api/dist/api/proto/daemon.js
"use strict";
// Code generated by protoc-gen-ts_proto. DO NOT EDIT.
// versions:
//   protoc-gen-ts_proto  v2.7.7
//   protoc               v6.32.0
// source: daemon.proto
Object.defineProperty(exports, "__esModule", { value: true });
exports.Response = exports.Request = exports.DmenuResponse = exports.DmenuRequest = exports.PingResponse = exports.PingRequest = exports.UrlRequest = exports.UrlResponse = exports.protobufPackage = void 0;
/* eslint-disable */
const wire_1 = require("@bufbuild/protobuf/wire");
exports.protobufPackage = "proto.ext.daemon";
function createBaseUrlResponse() {
    return {};
}
exports.UrlResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUrlResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.UrlResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBaseUrlResponse();
        return message;
    },
};
function createBaseUrlRequest() {
    return { url: "" };
}
exports.UrlRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.url !== "") {
            writer.uint32(10).string(message.url);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseUrlRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.url = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { url: isSet(object.url) ? globalThis.String(object.url) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.url !== "") {
            obj.url = message.url;
        }
        return obj;
    },
    create(base) {
        return exports.UrlRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseUrlRequest();
        message.url = object.url ?? "";
        return message;
    },
};
function createBasePingRequest() {
    return {};
}
exports.PingRequest = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePingRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.PingRequest.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBasePingRequest();
        return message;
    },
};
function createBasePingResponse() {
    return {};
}
exports.PingResponse = {
    encode(_, writer = new wire_1.BinaryWriter()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePingResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(_) {
        return {};
    },
    toJSON(_) {
        const obj = {};
        return obj;
    },
    create(base) {
        return exports.PingResponse.fromPartial(base ?? {});
    },
    fromPartial(_) {
        const message = createBasePingResponse();
        return message;
    },
};
function createBaseDmenuRequest() {
    return { rawContent: "", navigationTitle: "", placeholder: "", noIcon: false, sectionTitle: "", noSection: false };
}
exports.DmenuRequest = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.rawContent !== "") {
            writer.uint32(10).string(message.rawContent);
        }
        if (message.navigationTitle !== "") {
            writer.uint32(18).string(message.navigationTitle);
        }
        if (message.placeholder !== "") {
            writer.uint32(26).string(message.placeholder);
        }
        if (message.noIcon !== false) {
            writer.uint32(32).bool(message.noIcon);
        }
        if (message.sectionTitle !== "") {
            writer.uint32(42).string(message.sectionTitle);
        }
        if (message.noSection !== false) {
            writer.uint32(48).bool(message.noSection);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDmenuRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.rawContent = reader.string();
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.navigationTitle = reader.string();
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.placeholder = reader.string();
                    continue;
                }
                case 4: {
                    if (tag !== 32) {
                        break;
                    }
                    message.noIcon = reader.bool();
                    continue;
                }
                case 5: {
                    if (tag !== 42) {
                        break;
                    }
                    message.sectionTitle = reader.string();
                    continue;
                }
                case 6: {
                    if (tag !== 48) {
                        break;
                    }
                    message.noSection = reader.bool();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            rawContent: isSet(object.rawContent) ? globalThis.String(object.rawContent) : "",
            navigationTitle: isSet(object.navigationTitle) ? globalThis.String(object.navigationTitle) : "",
            placeholder: isSet(object.placeholder) ? globalThis.String(object.placeholder) : "",
            noIcon: isSet(object.noIcon) ? globalThis.Boolean(object.noIcon) : false,
            sectionTitle: isSet(object.sectionTitle) ? globalThis.String(object.sectionTitle) : "",
            noSection: isSet(object.noSection) ? globalThis.Boolean(object.noSection) : false,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.rawContent !== "") {
            obj.rawContent = message.rawContent;
        }
        if (message.navigationTitle !== "") {
            obj.navigationTitle = message.navigationTitle;
        }
        if (message.placeholder !== "") {
            obj.placeholder = message.placeholder;
        }
        if (message.noIcon !== false) {
            obj.noIcon = message.noIcon;
        }
        if (message.sectionTitle !== "") {
            obj.sectionTitle = message.sectionTitle;
        }
        if (message.noSection !== false) {
            obj.noSection = message.noSection;
        }
        return obj;
    },
    create(base) {
        return exports.DmenuRequest.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseDmenuRequest();
        message.rawContent = object.rawContent ?? "";
        message.navigationTitle = object.navigationTitle ?? "";
        message.placeholder = object.placeholder ?? "";
        message.noIcon = object.noIcon ?? false;
        message.sectionTitle = object.sectionTitle ?? "";
        message.noSection = object.noSection ?? false;
        return message;
    },
};
function createBaseDmenuResponse() {
    return { output: "" };
}
exports.DmenuResponse = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.output !== "") {
            writer.uint32(10).string(message.output);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDmenuResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.output = reader.string();
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return { output: isSet(object.output) ? globalThis.String(object.output) : "" };
    },
    toJSON(message) {
        const obj = {};
        if (message.output !== "") {
            obj.output = message.output;
        }
        return obj;
    },
    create(base) {
        return exports.DmenuResponse.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseDmenuResponse();
        message.output = object.output ?? "";
        return message;
    },
};
function createBaseRequest() {
    return { ping: undefined, url: undefined, dmenu: undefined };
}
exports.Request = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.ping !== undefined) {
            exports.PingRequest.encode(message.ping, writer.uint32(10).fork()).join();
        }
        if (message.url !== undefined) {
            exports.UrlRequest.encode(message.url, writer.uint32(18).fork()).join();
        }
        if (message.dmenu !== undefined) {
            exports.DmenuRequest.encode(message.dmenu, writer.uint32(26).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.ping = exports.PingRequest.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.url = exports.UrlRequest.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.dmenu = exports.DmenuRequest.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            ping: isSet(object.ping) ? exports.PingRequest.fromJSON(object.ping) : undefined,
            url: isSet(object.url) ? exports.UrlRequest.fromJSON(object.url) : undefined,
            dmenu: isSet(object.dmenu) ? exports.DmenuRequest.fromJSON(object.dmenu) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.ping !== undefined) {
            obj.ping = exports.PingRequest.toJSON(message.ping);
        }
        if (message.url !== undefined) {
            obj.url = exports.UrlRequest.toJSON(message.url);
        }
        if (message.dmenu !== undefined) {
            obj.dmenu = exports.DmenuRequest.toJSON(message.dmenu);
        }
        return obj;
    },
    create(base) {
        return exports.Request.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseRequest();
        message.ping = (object.ping !== undefined && object.ping !== null)
            ? exports.PingRequest.fromPartial(object.ping)
            : undefined;
        message.url = (object.url !== undefined && object.url !== null) ? exports.UrlRequest.fromPartial(object.url) : undefined;
        message.dmenu = (object.dmenu !== undefined && object.dmenu !== null)
            ? exports.DmenuRequest.fromPartial(object.dmenu)
            : undefined;
        return message;
    },
};
function createBaseResponse() {
    return { ping: undefined, url: undefined, dmenu: undefined };
}
exports.Response = {
    encode(message, writer = new wire_1.BinaryWriter()) {
        if (message.ping !== undefined) {
            exports.PingResponse.encode(message.ping, writer.uint32(10).fork()).join();
        }
        if (message.url !== undefined) {
            exports.UrlResponse.encode(message.url, writer.uint32(18).fork()).join();
        }
        if (message.dmenu !== undefined) {
            exports.DmenuResponse.encode(message.dmenu, writer.uint32(26).fork()).join();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof wire_1.BinaryReader ? input : new wire_1.BinaryReader(input);
        const end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1: {
                    if (tag !== 10) {
                        break;
                    }
                    message.ping = exports.PingResponse.decode(reader, reader.uint32());
                    continue;
                }
                case 2: {
                    if (tag !== 18) {
                        break;
                    }
                    message.url = exports.UrlResponse.decode(reader, reader.uint32());
                    continue;
                }
                case 3: {
                    if (tag !== 26) {
                        break;
                    }
                    message.dmenu = exports.DmenuResponse.decode(reader, reader.uint32());
                    continue;
                }
            }
            if ((tag & 7) === 4 || tag === 0) {
                break;
            }
            reader.skip(tag & 7);
        }
        return message;
    },
    fromJSON(object) {
        return {
            ping: isSet(object.ping) ? exports.PingResponse.fromJSON(object.ping) : undefined,
            url: isSet(object.url) ? exports.UrlResponse.fromJSON(object.url) : undefined,
            dmenu: isSet(object.dmenu) ? exports.DmenuResponse.fromJSON(object.dmenu) : undefined,
        };
    },
    toJSON(message) {
        const obj = {};
        if (message.ping !== undefined) {
            obj.ping = exports.PingResponse.toJSON(message.ping);
        }
        if (message.url !== undefined) {
            obj.url = exports.UrlResponse.toJSON(message.url);
        }
        if (message.dmenu !== undefined) {
            obj.dmenu = exports.DmenuResponse.toJSON(message.dmenu);
        }
        return obj;
    },
    create(base) {
        return exports.Response.fromPartial(base ?? {});
    },
    fromPartial(object) {
        const message = createBaseResponse();
        message.ping = (object.ping !== undefined && object.ping !== null)
            ? exports.PingResponse.fromPartial(object.ping)
            : undefined;
        message.url = (object.url !== undefined && object.url !== null) ? exports.UrlResponse.fromPartial(object.url) : undefined;
        message.dmenu = (object.dmenu !== undefined && object.dmenu !== null)
            ? exports.DmenuResponse.fromPartial(object.dmenu)
            : undefined;
        return message;
    },
};
function isSet(value) {
    return value !== null && value !== undefined;
}
